### DISCUSSION TITLE
Putting docker container behind a treafik reverse proxy

### DISCUSSION THREAD

**Discussion by phirestalker** [giongto35/cloud-game#399] [Q&A] [answered]
I have many services working on docker with traefik as my reverse proxy. I have tried to set this up pointing traefik at port 8000. I can get to the interface, but the games list is blank.

I am not sure where to look for errors. I have tried looking in docker logs for this container and for the traefik docker. I have also looked at the developer console in Chrome. I found not errors in any of them.

Has anyone gotten this to run behind a reverse proxy (hopefully traefik)?
Where else should I look for errors?

**Comment by sergystepanov** [comment]
If you look at the Docker Compose example:
https://github.com/giongto35/cloud-game/blob/167071af6f0c5c92e197afef7b754be1b58d6211/docker-compose.yml#L8-L19
In addition to one HTTP port 8000 of coordinator, there is one UDP port 8443 for worker (port 9000 doesn't matter).
WebRTC connection and streaming won't be possible without it, so make sure this additional port is reachable on your server.

All the logs should be visible in the console if `CLOUD_GAME_COORDINATOR_DEBUG` and `CLOUD_GAME_WORKER_DEBUG` environment variables are set to `true`. And with `PION_LOG_TRACE=all` it will show everything about WebRTC connections.

Also I added some error reporting when WebRTC ICE stuff fails after a 30 sec timeout. 
![image](https://github.com/giongto35/cloud-game/assets/846874/764cb401-706c-4616-b8fb-9c340c1c4d09)

**Reply by phirestalker** [reply]
I added an ingress for UDP port 8443, then I added a router and service to the docker-compose.yml file for the cloud-game container. I turned all those logging options on. The only error I can find is `sender_interceptor WARNING: 2023/05/17 14:25:17 failed sending: the DTLS transport has not started yet`. I do not see the error you mentioned.

Let me know what logs or config files you need.

**Reply by phirestalker** [reply]
Also, I am running this on my local network only. Do I need to remove the google stun server reference in the config?

**Reply by sergystepanov** [reply]
> I added an ingress for UDP port 8443, then I added a router and service to the docker-compose.yml file for the cloud-game container. I turned all those logging options on. The only error I can find is `sender_interceptor WARNING: 2023/05/17 14:25:17 failed sending: the DTLS transport has not started yet`. I do not see the error you mentioned.
> 
> Let me know what logs or config files you need.

Could you show me Docker logs, browser console output, and what's in the ICE candidates table on the tab in Chrome: chrome://webrtc-internals/ or Firefox: about:webrtc?

> Also, I am running this on my local network only. Do I need to remove the google stun server reference in the config?

STUN allows public IP detection and use, so it is not mandatory for a local network.

**Reply by phirestalker** [reply]
### web console log
[chrome console output.txt](https://github.com/giongto35/cloud-game/files/11501173/chrome.console.output.txt)
### docker logs
[docker_logs.txt](https://github.com/giongto35/cloud-game/files/11501186/docker_logs.txt)
### screenshot of table from webrtc-internals
<img width="1888" alt="Screenshot 2023-05-17 at 10 49 41 AM" src="https://github.com/giongto35/cloud-game/assets/7477095/1f27c827-8d21-49c3-b53b-d88fd0ac69af">
What is the correct environment variable to set public host/ip?

**Comment by sergystepanov** [comment]
Maybe try it with `CLOUD_GAME_WEBRTC_ICEIPMAP=127.0.0.1`? Your browser can't connect to 127.0.0.1:8443...

**Reply by phirestalker** [reply]
same result. Can I put a domain name in there?

EDIT:
nope, the ICE candidate table is empty with the FQDN put in there.

**Comment by sergystepanov** [comment]
> same result. Can I put a domain name in there?

IP adresses only.
Domains only work with TLS: 
```
worker/coordonator:
  server:
    tls:
      domain:
```

I wrote this thing some time ago https://github.com/sergystepanov/webrtc-troubleshooting. Basically, this is WebRTC part of the retro but   all those options can be changed on the fly.
Maybe try it? In Docker there are two ports 3000/TCP (for interface) and 4000/UDP (for WebRTC). And on the page just set "Use a single port" to 4000.

<details>
  <summary>w3t</summary>

![image](https://github.com/giongto35/cloud-game/assets/846874/68bc77e2-ebe2-4c00-822c-bd38327629de)

</details>

**Reply by phirestalker** [reply]
That gives me the error
`err: Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.`

EDIT:
sorry I mean the test server you linked to. I spun it up and when I click start, it gives me that error.

**Reply by sergystepanov** [reply]
Updated that w3t tool for WSS support. Tested it with NGINX reverse proxy, something like that:

```
# nginx.conf

http {

    server {
        listen 80;
        server_name site;

        location / {
            return 301 https://$host$request_uri;
        }
    }

    server {
        listen 443 ssl;
        server_name site;

        ssl_certificate /etc/ssl/certs/site.crt;
        ssl_certificate_key /etc/ssl/private/site.key;

        location / {
            proxy_pass         http://w3t-local:3000;
            proxy_http_version 1.1;
            proxy_set_header   Upgrade $http_upgrade;
            proxy_set_header Connection "Upgrade";
            proxy_set_header   Host $host;
        }
    }
}
```
```
# docker-compose.yml

version: '3'

services:
  proxy:
    image: nginx:latest
    ports:
      - "443:443"
      - "80:80"
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
      - ./nginx/cert/localhost.crt:/etc/ssl/certs/site.crt
      - ./nginx/cert/localhost.key:/etc/ssl/private/site.key

  w3t-local:
    build:
      context: .
    depends_on:
      - proxy
    ports:
      - "3000:3000"
      - "4000:4000/udp"
```

But I'm not really sure why do you even need that, since your problem somewhere in the UDP stuff, not it TCP (HTTP/S+WS/S). 🤷‍♂️

**Reply by phirestalker** [reply]
I could not continue with the test because it could not connect to the websocket. It stopped before trying any of the webrtc stuff.
I have tested with the new version. I created a new ingress for port 4000 and entered it in single port field. What are some things that I should change to track down the problem? Is the 1:1 NAT field where I put the IP similar to the ipmap setting in cloud-game?

**Reply by phirestalker** [reply]
Here are the logs. I have verified with multiple sites (including the traefik documentation) that I am setting up the udp port correctly. I am not sure what else it could be.
[w3t-log.txt](https://github.com/giongto35/cloud-game/files/11508281/w3t-log.txt)

**Reply by sergystepanov** [reply]
Set "Logging level (server)" to Trace, it will show ping results for client-server pairs.

**Reply by phirestalker** [reply]
[w3t-log (1).txt](https://github.com/giongto35/cloud-game/files/11508586/w3t-log.1.txt)

**Reply by sergystepanov** [reply]
Interesting. According to the logs, it can't reach your client IP from the server, which is your public IP.
And there is an ICE error: `Failure during addIceCandidate(): InvalidStateError`, don't know what it is.

> 2435454 s ice ping STUN from udp4 host 127.0.0.1:4000 to udp4 srflx ¯ \ _ (ツ) _ / ¯:58263 related 0.0.0.0:0
> 2435454 s ice ping STUN from udp4 host 10.254.0.23:4000 to udp4 srflx ¯ \ _ (ツ) _ / ¯:58263 related 0.0.0.0:0

So, your browser can only find your public `¯ \ _ (ツ) _ / ¯` IP and none of your local IPs.
I guess, one of the options is to find an IP for the server and set it with `CLOUD_GAME_WEBRTC_ICEIPMAP`, which will be accessible from your public IP (network), from the client. IDK.
And try it with w3t option: "Make server do offer".

I tried your server (from the Internet) and it worked for me, btw. (this is why there is STUN and TURN servers -- for mad NAT traversal, like in your setup).

**Reply by phirestalker** [reply]
The test succeeded when you ran it? If so it must be because the dns gives a local ip for the server in my LAN, but public IP outside of LAN. In that case, I should be able to put the server's LAN IP in CLOUD_GAME_WEBRTC_ICEIPMAP, but when I do that it still doesn't work.

EDIT:
can you post logs of your successful test?
This is crazy, there is no way it should work from the internet. I don't even have that port open on my firewall!!

**Reply by phirestalker** [reply]
Where is an equivalent option in w3t for CLOUD_GAME_WEBRTC_ICEIPMAP so I can test?

**Reply by sergystepanov** [reply]
It's "Use 1:1 NAT mapping" for ICEIPMAP.

But it works from the Internet somehow. my ip -- your ip.
>> 8666 c RTC done / [in-progress] [srflx] udp://¯ \ _ (ツ) _ / ¯:60300 --- [srflx] udp://¯ \ _ (ツ) _ / ¯:38050 (1513 ms)

<details>
<summary>log</summary>

```
7152 c RTC Start
7153 c WS connect to wss://webrtc.computeraccess.icu/websocket?ice_servers=stun%3Astun.nextcloud.com%3A443%2Cstun%3Astun.l.google.com%3A19302&log_level=4&ssl=true
7793 c WS → opened
7797 s sys log level is Debug
7798 c RTC SDP offer: v=0
o=- 5976405157811585382 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0
a=extmap-allow-mixed
a=msid-semantic: WMS
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=ice-ufrag:1o9p
a=ice-pwd:icbUl9K2NGshOFl9o9ZZZMV6
a=ice-options:trickle
a=fingerprint:sha-256 45:B7:91:D8:29:E0:9C:B6:C5:B4:6B:1D:4C:45:11:DB:69:CA:03:62:92:B0:F3:C3:67:3D:5D:48:36:AB:02:CB
a=setup:actpass
a=mid:0
a=sctp-port:5000
a=max-message-size:262144

7798 s sys secure? true
7799 c sig → have-local-offer
7799 c ICE → gathering
7819 c ICE local candidate:4098346846 1 udp 2113937151 eb94e84d-2418-46c4-93ec-fc6f2110e6e8.local 60300 typ host generation 0 ufrag 1o9p network-cost 999
7954 c ICE local candidate:2044234172 1 udp 1677729535 ¯ \ _ (ツ) _ / ¯ 60300 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag 1o9p network-cost 999
8010 s pc signaling state changed to have-remote-offer
8011 s sig → have-remote-offer
8011 s pc signaling state changed to stable
8011 s ice → gathering
8011 s ice Started agent: isControlling? false, remoteUfrag: "1o9p", remotePwd: "icbUl9K2NGshOFl9o9ZZZMV6"
8012 s ICE remote candidate:540462756 1 udp 2130706431 10.254.0.23 33108 typ host
8013 c sig → stable
8013 s RTC SDP answer: v=0
o=- 1483473032491247227 1684423857 IN IP4 0.0.0.0
s=-
t=0 0
a=fingerprint:sha-256 5F:61:6B:A4:B6:06:DA:3B:80:98:20:CD:2B:1C:42:48:8B:2F:BE:B6:0B:70:81:E3:50:B7:97:25:2D:60:2E:11
a=extmap-allow-mixed
a=group:BUNDLE 0
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=setup:active
a=mid:0
a=sendrecv
a=sctp-port:5000
a=ice-ufrag:kBbWVUTTfiuVZTRO
a=ice-pwd:memuvwAZwBNeIgoXVoaYpoKNTlkfzdeh

8013 c ICE → checking
8021 s ice Setting new connection state: Checking
8021 c RTC → connecting
8022 s pc ICE connection state changed: checking
8022 s ice → checking
8023 s pc peer connection state changed: connecting
8023 s rtc → connecting
8023 s ice Failed to ping without candidate pairs. Connection is not possible yet.
8023 s sig → stable
8053 s ice failed to get server reflexive address %!s(MISSING) %!s(MISSING): %!v(MISSING)
8220 s ice failed to get server reflexive address %!s(MISSING) %!s(MISSING): %!v(MISSING)
8221 s ICE remote candidate:35336446 1 udp 1694498815 ¯ \ _ (ツ) _ / ¯ 38050 typ srflx raddr 0.0.0.0 rport 38050
8221 s ice Failed to ping without candidate pairs. Connection is not possible yet.
8244 s ICE remote candidate:35336446 1 udp 1694498815 ¯ \ _ (ツ) _ / ¯ 45625 typ srflx raddr 0.0.0.0 rport 45625
8244 s ice → complete
8428 c ICE → complete
8429 c ICE → connected
8456 s ice Setting new connection state: Connected
8456 s pc ICE connection state changed: connected
8457 s ice → connected
8663 c RTC → connected
8666 c RTC done / [in-progress] [srflx] udp://¯ \ _ (ツ) _ / ¯:60300 --- [srflx] udp://¯ \ _ (ツ) _ / ¯:38050 (1513 ms)
8871 s pc peer connection state changed: connected
8872 s rtc → connected
8872 s sctp [0xc0002bc1e0] writeLoop entered
8872 s sctp [0xc0002bc1e0] readLoop entered
8873 s sctp [0xc0002bc1e0] state change: 'Closed' => 'CookieWait'
8875 s sctp [0xc0002bc1e0] sending INIT
8875 s sctp [0xc0002bc1e0] chunkInit received in state 'CookieWait'
8876 s sctp [0xc0002bc1e0] use ForwardTSN (on init)
9085 s sctp [0xc0002bc1e0] chunkInitAck received in state 'CookieWait'
9085 s sctp [0xc0002bc1e0] initial rwnd=5242880
9088 s sctp [0xc0002bc1e0] use ForwardTSN (on initAck)
9089 s sctp [0xc0002bc1e0] sending COOKIE-ECHO
9089 s sctp [0xc0002bc1e0] state change: 'CookieWait' => 'CookieEchoed'
9089 s sctp [0xc0002bc1e0] COOKIE-ECHO received in state 'CookieEchoed'
9089 s sctp [0xc0002bc1e0] state change: 'CookieEchoed' => 'Established'
9293 s sctp [0xc0002bc1e0] COOKIE-ACK received in state 'Established'
9293 s sctp [0xc0002bc1e0] accepted a new stream (streamIdentifier: 1)
9293 s sctp [1:0xc0002bc1e0] reassemblyQueue readable=true
9293 s sctp [1:0xc0002bc1e0] readNotifier.signal()
9294 s sctp [1:0xc0002bc1e0] readNotifier.signal() done
9294 s RTC ○○○○●●○●
9296 s sctp [1:0xc0002bc1e0] reliability params: ordered=true type=0 value=0
9493 s sctp [0xc0002bc1e0] sending SACK: SACK cumTsnAck=2161656232 arwnd=1048576 dupTsn=[]
18510 s sctp [0xc0002bc1e0] resetStream(): senderLastTSN=2161656232 <= peerLastTSN=2161656232
18511 s sctp [1:0xc0002bc1e0] onInboundStreamReset: state=open
18511 s sctp [0xc0002bc1e0] deleting stream 1
18511 s sctp [0xc0002bc1e0] closing association..
18511 s sctp [0xc0002bc1e0] state change: 'Established' => 'Closed'
18512 s sctp [1:0xc0002bc1e0] Close: state=open
18512 s sctp [1:0xc0002bc1e0] state change: open => closed
18512 s sctp [0xc0002bc1e0] readLoop exited [0xc0002bc1e0] abort chunk, with following errors: (User Initiated Abort: Close called)
18512 s sctp [0xc0002bc1e0] association closed
18512 s sctp [0xc0002bc1e0] stats nDATAs (in) : 1
18720 s sctp [0xc0002bc1e0] failed to write packets on netConn: dtls fatal: conn is closed
18720 s sctp [0xc0002bc1e0] writeLoop ended
18720 s sctp [0xc0002bc1e0] writeLoop exited
18720 s sctp [0xc0002bc1e0] stats nSACKs (in) : 2
18721 s sig !close
18721 s sctp [0xc0002bc1e0] stats nT3Timeouts : 0
18721 s sctp [0xc0002bc1e0] closing association..
18721 s sctp [0xc0002bc1e0] closing association..
18722 s sctp [0xc0002bc1e0] association closed
18722 s sctp [0xc0002bc1e0] stats nDATAs (in) : 1
18722 s sctp [0xc0002bc1e0] stats nSACKs (in) : 2
18722 s sctp [0xc0002bc1e0] stats nT3Timeouts : 0
18723 s sctp [0xc0002bc1e0] stats nAckTimeouts: 1
18723 s sctp [0xc0002bc1e0] stats nAckTimeouts: 1
18723 s sctp [0xc0002bc1e0] stats nFastRetrans: 0
18723 s sctp [0xc0002bc1e0] stats nFastRetrans: 0
18724 s pc Failed to accept RTP stream is already closed
18724 s pc Failed to accept RTCP stream is already closed
18724 s ice Failed to discover mDNS candidate eb94e84d-2418-46c4-93ec-fc6f2110e6e8.local: mDNS: connection is closed
18724 s ice Setting new connection state: Closed
18725 s pc peer connection state changed: closed
18931 c WS → closed: 1000 (Normal Closure)
18932 c RTC Stop
```
</details>

**Reply by phirestalker** [reply]
Thanks. Can you tell me the format of the 1:1 map setting? I have tried just putting an IP in there, but that is not correct.

**Reply by sergystepanov** [reply]
I think, x.x.x.x/host

More precise:
	/host
	/srflx
	/prflx
	/relay
Try /host and /srflx.

(I need to add that into the description for that option. 😅)

**Reply by phirestalker** [reply]
I got a successful test with server offer and the local ip of the server /host. However, I can not seem to get cloud-game to connect to its webrtc port. I set ip map with the local server ip, but it still is unable to connect. I will keep trying different things. Is there a way to make cloud-game server do the offer?

EDIT:
uuuugh, nevermind. The test succeeds without make server do offer, so that is not it.

**Reply by phirestalker** [reply]
If it is always relying on a stun server, then it will never connect when served on my local LAN. Should I try removing the stun server from the config file, or should I spin one up in the docker-compose and use that?

**Reply by sergystepanov** [reply]
Try disabling this feature in Chrome: chrome://flags/#enable-webrtc-hide-local-ips-with-mdns. This will allow you to see the IP address being resolved in the browser, instead of obfuscated xxxxxx.local addresses. If this IP is unreachable from your server, then, I think, local STUN may help...

**Reply by phirestalker** [reply]
Will do. When I enter an IP address for ICEIPMAP, what is it using internally for the type? Is it host? If not, maybe allowing a setting for that will help.

EDIT:
oh, and it is reachable from my browser. My concern was that the STUN servers can not see it because it is a local LAN address. Your, and I think, my successful connection with w3t was using some public address figured out by the STUN server. It was using a port that was not 4000. If it would just try to connect locally without the STUN server (if that's even possible), then it should work.

**Reply by sergystepanov** [reply]
> When I enter an IP address for ICEIPMAP, what is it using internally for the type? Is it host?

Yes, it doesn't matter.

> I got a successful test with server offer and the local ip of the server /host. However, I can not seem to get cloud-game to connect to its webrtc port. I set ip map with the local server ip, but it still is unable to connect. I will keep trying different things. Is there a way to make cloud-game server do the offer?

In cloud-game it's the server who makes an offer (for historical reasons due to old Firefox problems).
Strange. It should work similar to w3t.

I'm out of ideas. Last thing which you could try, is to run Docker in your Host network dropping all explicit port mappings. Like we do in cloudretro.io.
https://github.com/giongto35/cloud-game/blob/624eecd4e8ef3a26dabd3c51afc9793b01fba26c/.github/workflows/cd/docker-compose.yml#L6-L7
The idea is that you can leave a single port for WebRTC and proxy it trough your proxy, or just leave your UDPs outta Traefik in singleport/auto-port modes, removing one layer of networking (Docker bridge). ?

**Reply by phirestalker** [reply]
I got it working. I have the local IP in ICEIPMAP, but I forgot to forward the udp port. I kept forwarding and not forwarding for each thing I tried. I must have forgotten to forward it again for the last tweak.

I see a few issue with the few minutes of gameplay on super mario bros. Should I discuss them here?

**Reply by sergystepanov** [reply]
Cool. 
Would be better not to discuss unrelated issues in a single discussion. Also we have Discord.
Just add a brief comment about how you solved it as a reply on the last message (https://github.com/giongto35/cloud-game/discussions/399#discussioncomment-5931369) in this discussion + mark it as answer, that would help others.

**Comment by sergystepanov** [comment]
In short, I assume, there is a problem in your UDP ingress config what makes this port (or IP) unreachable from the client. And the only theoretically useful option here is `CLOUD_GAME_WEBRTC_ICEIPMAP`, which forces to override, I think all, IP-addresses announced by the worker. This is useful in cases, for example, if worker inside Docker gathers only internal IPs 172.17.x.x and such.
Try to look with that tool what IPs:ports it gathers and what's up with their reachability (set Trace level instead of Debug in the options on the page for that). 🤷‍♂️

**Reply by phirestalker** [reply, marked as answer]
I was able to solve this by using ICEIPMAP of my server's LAN address, and forwarding the UDP port in the docker compose file. I think that is all. I made a lot of back and forth changes.
